File: autocontext/src/autocontext/session/background_session_read_model.py
from __future__ import annotations

import json
from collections.abc import Mapping, Sequence
from typing import Any, Literal, TypeAlias
from urllib.parse import quote

from autocontext.session.runtime_events import RuntimeSessionEventLog

BackgroundSessionStatus: TypeAlias = Literal["queued", "running", "completed", "failed", "canceled", "skipped", "unknown"]
BackgroundSessionSummary: TypeAlias = dict[str, str | int]
BackgroundSessionArtifact: TypeAlias = dict[str, str]
BackgroundSessionDetail: TypeAlias = dict[str, Any]
TaskQueueRowLike: TypeAlias = Mapping[str, Any]
RunRowLike: TypeAlias = Mapping[str, Any]
ArtifactLike: TypeAlias = Mapping[str, Any]

REDACTED_TRIGGER_VALUE = "[redacted]"
_SENSITIVE_TRIGGER_KEY_WORDS = frozenset(
    {
        "auth",
        "apikey",
        "authorization",
        "bearer",
        "credential",
        "credentials",
        "password",
        "passwd",
        "privatekey",
        "secret",
        "token",
    }
)
_COMPOUND_SENSITIVE_TRIGGER_KEY_WORDS = (
    ("api", "key"),
    ("private", "key"),
    ("access", "token"),
    ("refresh", "token"),
)
_SECRET_VALUE_MARKERS = (
    "ghp_",
    "gho_",
    "ghu_",
    "ghs_",
    "ghr_",
    "github_pat_",
    "sk-",
    "bearer ",
)


def build_background_session_summary(
    *,
    runtime_session: RuntimeSessionEventLog | None = None,
    task: TaskQueueRowLike | None = None,
    run: RunRowLike | None = None,
    artifacts: Sequence[ArtifactLike] | None = None,
    child_sessions: Sequence[RuntimeSessionEventLog] | None = None,
) -> BackgroundSessionSummary:
    artifacts = artifacts or []
    child_sessions = child_sessions or []
    session_id = runtime_session.session_id if runtime_session else _task_session_id(task)
    created_at = (
        runtime_session.created_at if runtime_session else _task_str(task, "created_at") or _run_str(run, "created_at")
    )
    updated_at = _updated_at(runtime_session, task, run)
    status = _normalize_status(
        _metadata_str(runtime_session, "status") or _task_str(task, "status") or _run_str(run, "status"),
        has_runtime_session=runtime_session is not None,
    )

    return {
        "session_id": session_id,
        "runtime_session_id": runtime_session.session_id if runtime_session else "",
        "run_id": _metadata_str(runtime_session, "runId") or _run_str(run, "run_id"),
        "task_id": runtime_session.task_id if runtime_session and runtime_session.task_id else _task_str(task, "id"),
        "parent_session_id": runtime_session.parent_session_id if runtime_session else "",
        "status": status,
        "goal": _metadata_str(runtime_session, "goal") or _task_str(task, "spec_name") or _run_str(run, "scenario"),
        "event_count": len(runtime_session.events) if runtime_session else 0,
        "artifact_count": len(artifacts),
        "child_session_count": len(child_sessions),
        "created_at": created_at,
        "updated_at": updated_at,
        "result_url": background_session_url(session_id),
        "runtime_session_url": runtime_session_url(runtime_session.session_id if runtime_session else ""),
    }


def build_background_session_detail(
    *,
    runtime_session: RuntimeSessionEventLog | None = None,
    task: TaskQueueRowLike | None = None,
    run: RunRowLike | None = None,
    artifacts: Sequence[ArtifactLike] | None = None,
    child_sessions: Sequence[RuntimeSessionEventLog] | None = None,
) -> BackgroundSessionDetail:
    return {
        "summary": build_background_session_summary(
            runtime_session=runtime_session,
            task=task,
            run=run,
            artifacts=artifacts,
            child_sessions=child_sessions,
        ),
        "artifacts": [_sanitize_artifact(artifact) for artifact in (artifacts or [])],
        "child_sessions": [build_background_session_summary(runtime_session=child) for child in (child_sessions or [])],
        "trigger": _read_trigger(task),
    }


def background_session_url(session_id: str) -> str:
    return f"/api/cockpit/background-sessions/{quote(session_id, safe='')}" if session_id else ""


def runtime_session_url(session_id: str) -> str:
    return f"/api/cockpit/runtime-sessions/{quote(session_id, safe='')}" if session_id else ""


def _sanitize_artifact(artifact: ArtifactLike) -> BackgroundSessionArtifact:
    return {
        "artifact_id": _read_str(artifact.get("artifact_id")),
        "kind": _read_str(artifact.get("kind")) or "file",
        "label": _read_str(artifact.get("label")),
        "path": _read_str(artifact.get("path")),
        "url": _read_str(artifact.get("url")),
    }


def _read_trigger(task: TaskQueueRowLike | None) -> dict[str, str | int | bool] | None:
    raw = _task_str(task, "config_json")
    if not raw:
        return None
    try:
        parsed = json.loads(raw)
    except json.JSONDecodeError:
        return None
    if not isinstance(parsed, dict):
        return None
    trigger = parsed.get("trigger")
    if not isinstance(trigger, dict):
        return None
    return _sanitize_record(trigger)


def _sanitize_record(record: Mapping[str, Any]) -> dict[str, str | int | bool]:
    clean: dict[str, str | int | bool] = {}
    for key, value in record.items():
        if isinstance(key, str) and isinstance(value, str | int | bool):
            clean[key] = REDACTED_TRIGGER_VALUE if _is_sensitive_trigger_entry(key, value) else value
    return clean


def _is_sensitive_trigger_entry(key: str, value: str | int | bool) -> bool:
    return _is_sensitive_trigger_key(key) or (isinstance(value, str) and _looks_like_secret_value(value))


def _is_sensitive_trigger_key(key: str) -> bool:
    words = _trigger_key_words(key)
    if any(word in _SENSITIVE_TRIGGER_KEY_WORDS for word in words):
        return True
    return any(_contains_word_sequence(words, sequence) for sequence in _COMPOUND_SENSITIVE_TRIGGER_KEY_WORDS)


def _trigger_key_words(key: str) -> list[str]:
    words: list[str] = []
    current: list[str] = []
    previous = ""
    for character in key:
        if not character.isalnum():
            _append_trigger_key_word(words, current)
        elif current and character.isupper() and (previous.islower() or previous.isdigit()):
            _append_trigger_key_word(words, current)
            current.append(character)
        else:
            current.append(character)
        previous = character
    _append_trigger_key_word(words, current)
    return words


def _append_trigger_key_word(words: list[str], current: list[str]) -> None:
    if current:
        words.append("".join(current).lower())
        current.clear()


def _contains_word_sequence(words: Sequence[str], sequence: Sequence[str]) -> bool:
    if len(sequence) > len(words):
        return False
    end = len(words) - len(sequence) + 1
    for index in range(end):
        if list(words[index : index + len(sequence)]) == list(sequence):
            return True
    return False


def _looks_like_secret_value(value: str) -> bool:
    lower_value = value.lower()
    return any(marker in lower_value for marker in _SECRET_VALUE_MARKERS) or (
        "-----begin " in lower_value and "private key-----" in lower_value
    )


def _normalize_status(raw: str, *, has_runtime_session: bool) -> BackgroundSessionStatus:
    status = raw.strip().lower().replace("-", "_")
    if status in {"pending", "queued", "scheduled", "backlog"}:
        return "queued"
    if status in {"running", "started", "in_progress", "processing"}:
        return "running"
    if status in {"completed", "complete", "done", "success", "succeeded"}:
        return "completed"
    if status in {"failed", "failure", "error"}:
        return "failed"
    if status in {"canceled", "cancelled"}:
        return "canceled"
    if status == "skipped":
        return "skipped"
    return "running" if has_runtime_session else "unknown"


def _updated_at(
    runtime_session: RuntimeSessionEventLog | None,
    task: TaskQueueRowLike | None,
    run: RunRowLike | None,
) -> str:
    if runtime_session:
        return runtime_session.updated_at or runtime_session.created_at
    return (
        _task_str(task, "updated_at")
        or _task_str(task, "created_at")
        or _run_str(run, "updated_at")
        or _run_str(run, "created_at")
    )


def _task_session_id(task: TaskQueueRowLike | None) -> str:
    task_id = _task_str(task, "id")
    return f"task:{task_id}" if task_id else ""


def _metadata_str(runtime_session: RuntimeSessionEventLog | None, key: str) -> str:
    value = runtime_session.metadata.get(key) if runtime_session else None
    return value if isinstance(value, str) else ""


def _task_str(task: TaskQueueRowLike | None, key: str) -> str:
    value = task.get(key) if task else None
    return value if isinstance(value, str) else ""


def _run_str(run: RunRowLike | None, key: str) -> str:
    value = run.get(key) if run else None
    return value if isinstance(value, str) else ""


def _read_str(value: Any) -> str:
    return value if isinstance(value, str) else ""

File: ts/src/session/background-session-read-model.ts
import type { RuntimeSessionEventLog } from "./runtime-events.js";
import type { RunRow, TaskQueueRow } from "../storage/index.js";

export type BackgroundSessionStatus =
  | "queued"
  | "running"
  | "completed"
  | "failed"
  | "canceled"
  | "skipped"
  | "unknown";

export type BackgroundSessionArtifactKind =
  | "branch"
  | "commit"
  | "pull_request"
  | "screenshot"
  | "report"
  | "trace"
  | "dataset"
  | "verification_result"
  | "file"
  | string;

export interface BackgroundSessionArtifactInput {
  readonly artifact_id?: string;
  readonly kind?: BackgroundSessionArtifactKind;
  readonly label?: string;
  readonly path?: string;
  readonly url?: string;
  readonly metadata?: Record<string, unknown>;
}

export interface BackgroundSessionArtifact {
  readonly artifact_id: string;
  readonly kind: BackgroundSessionArtifactKind;
  readonly label: string;
  readonly path: string;
  readonly url: string;
}

export interface BackgroundSessionSummary {
  readonly session_id: string;
  readonly runtime_session_id: string;
  readonly run_id: string;
  readonly task_id: string;
  readonly parent_session_id: string;
  readonly status: BackgroundSessionStatus;
  readonly goal: string;
  readonly event_count: number;
  readonly artifact_count: number;
  readonly child_session_count: number;
  readonly created_at: string;
  readonly updated_at: string;
  readonly result_url: string;
  readonly runtime_session_url: string;
}

export interface BackgroundSessionDetail {
  readonly summary: BackgroundSessionSummary;
  readonly artifacts: BackgroundSessionArtifact[];
  readonly child_sessions: BackgroundSessionSummary[];
  readonly trigger: Record<string, unknown> | null;
}

export interface BackgroundSessionSource {
  readonly runtimeSession?: RuntimeSessionEventLog | null;
  readonly task?: TaskQueueRow | null;
  readonly run?: RunRow | null;
  readonly artifacts?: BackgroundSessionArtifactInput[] | null;
  readonly childSessions?: RuntimeSessionEventLog[] | null;
}

type NormalizedBackgroundSessionSource = {
  runtimeSession: RuntimeSessionEventLog | null;
  task: TaskQueueRow | null;
  run: RunRow | null;
  artifacts: BackgroundSessionArtifactInput[];
  childSessions: RuntimeSessionEventLog[];
};

const REDACTED_TRIGGER_VALUE = "[redacted]";
const SENSITIVE_TRIGGER_KEY_WORDS = new Set([
  "auth",
  "authorization",
  "bearer",
  "credential",
  "credentials",
  "password",
  "passwd",
  "secret",
  "token",
]);
const COMPOUND_SENSITIVE_TRIGGER_KEY_WORDS = [
  ["api", "key"],
  ["private", "key"],
  ["access", "token"],
  ["refresh", "token"],
] as const;
const SECRET_VALUE_PATTERNS = [
  /gh[pousr]_[A-Za-z0-9_]+/,
  /github_pat_[A-Za-z0-9_]+/,
  /\bsk-[A-Za-z0-9_-]{6,}\b/,
  /\bbearer\s+[A-Za-z0-9._~+/=-]+/i,
  /-----BEGIN [A-Z ]*PRIVATE KEY-----/,
] as const;
const CAMEL_CASE_BOUNDARY_PATTERN = /([a-z0-9])([A-Z])/g;
const TRIGGER_KEY_WORD_PATTERN = /[^A-Za-z0-9]+/;

export function buildBackgroundSessionSummary(
  source: BackgroundSessionSource,
): BackgroundSessionSummary {
  const normalized = normalizeSource(source);
  const sessionId = readSessionId(normalized);
  const runtimeSessionId = normalized.runtimeSession?.sessionId ?? "";

  return {
    session_id: sessionId,
    runtime_session_id: runtimeSessionId,
    run_id: readMetadataString(normalized.runtimeSession?.metadata, "runId") || normalized.run?.run_id || "",
    task_id: readTaskId(normalized),
    parent_session_id: normalized.runtimeSession?.parentSessionId ?? "",
    status: readBackgroundSessionStatus(normalized),
    goal: readGoal(normalized),
    event_count: normalized.runtimeSession?.events.length ?? 0,
    artifact_count: normalized.artifacts.length,
    child_session_count: normalized.childSessions.length,
    created_at: readCreatedAt(normalized),
    updated_at: readUpdatedAt(normalized),
    result_url: backgroundSessionUrl(sessionId),
    runtime_session_url: runtimeSessionUrl(runtimeSessionId),
  };
}

export function buildBackgroundSessionDetail(
  source: BackgroundSessionSource,
): BackgroundSessionDetail {
  return {
    summary: buildBackgroundSessionSummary(source),
    artifacts: (source.artifacts ?? []).map(sanitizeArtifact),
    child_sessions: (source.childSessions ?? []).map((runtimeSession) =>
      buildBackgroundSessionSummary({ runtimeSession }),
    ),
    trigger: readTrigger(source.task),
  };
}

export function backgroundSessionUrl(sessionId: string): string {
  return sessionId ? `/api/cockpit/background-sessions/${encodeURIComponent(sessionId)}` : "";
}

export function runtimeSessionUrl(sessionId: string): string {
  return sessionId ? `/api/cockpit/runtime-sessions/${encodeURIComponent(sessionId)}` : "";
}

const QUEUED_STATUSES = new Set(["pending", "queued", "scheduled", "backlog"]);
const RUNNING_STATUSES = new Set(["running", "started", "in_progress", "processing"]);
const COMPLETED_STATUSES = new Set(["completed", "complete", "done", "success", "succeeded"]);
const FAILED_STATUSES = new Set(["failed", "failure", "error"]);
const CANCELED_STATUSES = new Set(["canceled", "cancelled"]);

function normalizeSource(source: BackgroundSessionSource): NormalizedBackgroundSessionSource {
  return {
    runtimeSession: source.runtimeSession ?? null,
    task: source.task ?? null,
    run: source.run ?? null,
    artifacts: source.artifacts ?? [],
    childSessions: source.childSessions ?? [],
  };
}

function readSessionId(source: NormalizedBackgroundSessionSource): string {
  return source.runtimeSession?.sessionId || (source.task?.id ? `task:${source.task.id}` : "");
}

function readTaskId(source: NormalizedBackgroundSessionSource): string {
  return source.runtimeSession?.taskId || source.task?.id || "";
}

function readBackgroundSessionStatus(
  source: NormalizedBackgroundSessionSource,
): BackgroundSessionStatus {
  return normalizeStatus(
    readMetadataString(source.runtimeSession?.metadata, "status") ||
      source.task?.status ||
      source.run?.status ||
      "",
    Boolean(source.runtimeSession),
  );
}

function readGoal(source: NormalizedBackgroundSessionSource): string {
  return (
    readMetadataString(source.runtimeSession?.metadata, "goal") ||
    source.task?.spec_name ||
    source.run?.scenario ||
    ""
  );
}

function readCreatedAt(source: NormalizedBackgroundSessionSource): string {
  return source.runtimeSession?.createdAt || source.task?.created_at || source.run?.created_at || "";
}

function readUpdatedAt(source: NormalizedBackgroundSessionSource): string {
  return source.runtimeSession
    ? source.runtimeSession.updatedAt || source.runtimeSession.createdAt
    : source.task?.updated_at ||
        source.task?.created_at ||
        source.run?.updated_at ||
        source.run?.created_at ||
        "";
}

function sanitizeArtifact(artifact: BackgroundSessionArtifactInput): BackgroundSessionArtifact {
  return {
    artifact_id: artifact.artifact_id ?? "",
    kind: artifact.kind ?? "file",
    label: artifact.label ?? "",
    path: artifact.path ?? "",
    url: artifact.url ?? "",
  };
}

function readTrigger(task: TaskQueueRow | null | undefined): Record<string, unknown> | null {
  if (!task?.config_json) return null;
  const parsed = parseRecordJson(task.config_json);
  const trigger = parsed.trigger;
  if (!isRecord(trigger)) return null;
  return sanitizeRecord(trigger);
}

function normalizeStatus(raw: string, hasRuntimeSession: boolean): BackgroundSessionStatus {
  const status = raw.trim().toLowerCase().replace(/-/g, "_");
  if (QUEUED_STATUSES.has(status)) return "queued";
  if (RUNNING_STATUSES.has(status)) return "running";
  if (COMPLETED_STATUSES.has(status)) return "completed";
  if (FAILED_STATUSES.has(status)) return "failed";
  if (CANCELED_STATUSES.has(status)) return "canceled";
  if (status === "skipped") return "skipped";
  return hasRuntimeSession ? "running" : "unknown";
}

function parseRecordJson(json: string): Record<string, unknown> {
  try {
    const parsed: unknown = JSON.parse(json);
    return isRecord(parsed) ? parsed : {};
  } catch {
    return {};
  }
}

function sanitizeRecord(record: Record<string, unknown>): Record<string, unknown> {
  const clean: Record<string, unknown> = {};
  for (const [key, value] of Object.entries(record)) {
    if (isScalarTriggerValue(value)) {
      clean[key] = isSensitiveTriggerEntry(key, value) ? REDACTED_TRIGGER_VALUE : value;
    }
  }
  return clean;
}

function isScalarTriggerValue(value: unknown): value is string | number | boolean {
  return typeof value === "string" || typeof value === "number" || typeof value === "boolean";
}

function isSensitiveTriggerEntry(key: string, value: string | number | boolean): boolean {
  return isSensitiveTriggerKey(key) || (typeof value === "string" && looksLikeSecretValue(value));
}

function isSensitiveTriggerKey(key: string): boolean {
  const words = triggerKeyWords(key);
  if (words.some((word) => SENSITIVE_TRIGGER_KEY_WORDS.has(word))) {
    return true;
  }
  return COMPOUND_SENSITIVE_TRIGGER_KEY_WORDS.some((sequence) => containsWordSequence(words, sequence));
}

function triggerKeyWords(key: string): string[] {
  return key
    .replace(CAMEL_CASE_BOUNDARY_PATTERN, "$1 $2")
    .split(TRIGGER_KEY_WORD_PATTERN)
    .filter(Boolean)
    .map((word) => word.toLowerCase());
}

function containsWordSequence(words: readonly string[], sequence: readonly string[]): boolean {
  if (sequence.length > words.length) {
    return false;
  }
  const lastStart = words.length - sequence.length;
  for (let index = 0; index <= lastStart; index += 1) {
    let matches = true;
    for (let offset = 0; offset < sequence.length; offset += 1) {
      if (words[index + offset] !== sequence[offset]) {
        matches = false;
        break;
      }
    }
    if (matches) {
      return true;
    }
  }
  return false;
}

function looksLikeSecretValue(value: string): boolean {
  return SECRET_VALUE_PATTERNS.some((pattern) => pattern.test(value));
}

function readMetadataString(
  metadata: Record<string, unknown> | null | undefined,
  key: string,
): string {
  const value = metadata?.[key];
  return typeof value === "string" ? value : "";
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}
